File: ShoppingV5/app/src/main/java/dk/itu/shoppingv5/ItemsDB.java
package dk.itu.shoppingv5;

import java.util.HashMap;

public class ItemsDB {
    private static ItemsDB sItemsDB;
    private final HashMap<String, String> itemsMap = new HashMap<String, String>();

    private ItemsDB() {
        fillItemsDB();
    }

    public static ItemsDB get() {
        if (sItemsDB == null) sItemsDB = new ItemsDB();
        return sItemsDB;
    }

    public HashMap<String, String> getItemsDB() {
        return itemsMap;
    }

    public String listItems() {
        String r = "";
        for (HashMap.Entry<String, String> item : itemsMap.entrySet())
            r = r + "\n Buy " + item.getKey() + " in: " + item.getValue();
        return r;
    }

    public int size() {
        return itemsMap.size();
    }

    public String getWhere(String what) {
        return itemsMap.get(what);
    }

    public void addItem(String what, String where) {
        itemsMap.put(what, where);
    }

    public void removeItem(String what) {
        itemsMap.remove(what);
    }

    public boolean isPresent(String what) {
        return itemsMap.get(what) != null;
    }

    private void fillItemsDB() {
        itemsMap.put("coffee", "Irma");
        itemsMap.put("carrots", "Netto");
        itemsMap.put("milk", "Netto");
        itemsMap.put("bread", "bakery");
        itemsMap.put("butter", "Irma");
    }
}

File: ShoppingV4Kotlin/app/src/main/java/dk/itu/shoppingv4/ItemsDB.kt
package dk.itu.shoppingv4

class ItemsDB private constructor() {
    private val itemsDB = HashMap<String, String>()

    init {
        fillItemsDB()
    }

    companion object {
        private var sItemsDB: ItemsDB? = null

        fun get(): ItemsDB {
            return sItemsDB ?: ItemsDB().also{sItemsDB = it}
        }
    }

    fun listItems(): String {
        var r = ""
        for ((key, value) in itemsDB)
            r = "$r\n Buy $key in: $value"
        return r
    }

    fun size(): Int {
        return itemsDB.size
    }

    fun getWhere(what: String): String? {
        return itemsDB[what]
    }

    fun addItem(what: String, where: String) {
        itemsDB[what] = where
    }

    fun removeItem(what: String) {
        if (itemsDB[what] != null) itemsDB.remove(what)
    }

    private fun fillItemsDB() {
        itemsDB["coffee"] = "Irma"
        itemsDB["carrots"] = "Netto"
        itemsDB["milk"] = "Netto"
        itemsDB["bread"] = "bakery"
        itemsDB["butter"] = "Irma"
    }
}